File: javascript/ch3/sort.js
import _ from 'lodash';

const names = [
  'alonzo church',
  'Haskell curry',
  'stephen_kleene',
  'John Von Neumann',
  'stephen_kleene'
];

function sort_ip() {
  const result = [];
  for (let i = 0; i < names.length; i++) {
    const n = names[i];
    if (n != null) {
      const ns = n.replace(/_/, ' ').split(' ');
      for (let j = 0; j < ns.length; j++) {
        const p = ns[j];
        ns[j] = p.charAt(0).toUpperCase() + p.slice(1);
      }

      if (result.indexOf(ns.join(' ')) < 0) {
        result.push(ns.join(' '));
      }
    }
  }

  result.sort();

  console.log('sort_ip', result);
}

function sort_fp() {
  const result = _.chain(names)
    .filter((name) => !_.isUndefined(name))
    .map((name) => name.replace(/_/, ' '))
    .uniq()
    .map(_.startCase)
    .sort()
    .value();

  console.log('sort_fp', result);
}

export {
  sort_ip,
  sort_fp
};

File: python/python/ch3/sort.py
import toolz as tz
import pydash

names = [
    'alonzo church',
    'Haskell curry',
    'stephen_kleene',
    'John Von Neumann',
    'stephen_kleene'
]

def sort_ip():
    result = []
    for n in names:
        if n != None:
            ns = n.replace('_', ' ')
            ns = ns.split(' ')
            for j in range(len(ns)):
                p = ns[j]
                ns[j] = p[0].upper() + p[1:]
                
            if ' '.join(ns) not in result:
                result.append(' '.join(ns))
                
    result.sort()
    
    print('sort_ip', result)
  
def sort_fp():
    result = (pydash.chain(names)
              .filter_(lambda name: name != None)
              .map_(lambda name: name.replace('_', ' '))
              .uniq()
              .map(pydash.start_case)
              .sort()
              .value())
    
    print('sort_fp', result)
    
sort = [
    sort_ip,
    sort_fp
]